ghcide/src/Development/IDE/Core/PluginUtils.hs

```diff
@@ -0,0 +1,108 @@
+    RuleFailed rule       -> "RuleFailed:" <+> viaShow rule
+    CoreError perror      -> pretty $ PluginUtils.prettyPluginError perror
+
+pluginResponse ::
+    Monad m =>
+    ExceptT GhcidePluginError m a ->
+    m (Either LSP.ResponseError a)
+pluginResponse = PluginUtils.pluginResponse' handlePluginError
+
+withPluginError :: Functor m => ExceptT PluginUtils.PluginError m a -> ExceptT GhcidePluginError m a
+withPluginError = PluginUtils.withError CoreError
+
+mkPluginErrorMessage :: T.Text -> GhcidePluginError
+mkPluginErrorMessage = CoreError . PluginUtils.mkPluginErrorMessage
+
+handlePluginError :: GhcidePluginError -> LSP.ResponseError
+handlePluginError msg = PluginUtils.mkSimpleResponseError $ renderStrict simpleDoc
+    where
+        simpleDoc = layoutPretty defaultLayoutOptions $ pretty msg
+
+-- ----------------------------------------------------------------------------
+-- Action wrappers
+-- ----------------------------------------------------------------------------
+
+runAction :: MonadIO m => String -> IdeState -> ExceptT e Action a -> ExceptT e m a
+runAction herald ide act =
+  PluginUtils.hoistExceptT . ExceptT $
+    join $ shakeEnqueue (shakeExtras ide) (mkDelayedAction herald Logger.Debug $ runExceptT act)
+
+-- | Request a Rule result, it not available return the last computed result which may be stale.
+--   Errors out if none available.
+useWithStale_ ::(IdeRule k v)
+    => k -> NormalizedFilePath -> ExceptT e Action (v, PositionMapping)
+useWithStale_ key file = ExceptT $ fmap Right $ Shake.useWithStale_ key file
+
+useWithStale :: IdeRule k v
+    => k -> NormalizedFilePath -> ExceptT GhcidePluginError Action (v, PositionMapping)
+useWithStale key file = maybeToExceptT (FastRuleNotReady key) $ useWithStaleMaybeT key file
+
+-- | useE is useful to implement functions that aren’t rules but need shortcircuiting
+-- e.g. getDefinition.
+use :: IdeRule k v => k -> NormalizedFilePath -> ExceptT GhcidePluginError Action v
+use k = maybeToExceptT (RuleFailed k) . MaybeT . Shake.use k
+
+useWithStaleMaybeT :: IdeRule k v
+    => k -> NormalizedFilePath -> MaybeT Action (v, PositionMapping)
+useWithStaleMaybeT key file = MaybeT $ runIdentity <$> Shake.usesWithStale key (Identity file)
+
+-- ----------------------------------------------------------------------------
+-- IdeAction wrappers
+-- ----------------------------------------------------------------------------
+
+runIdeAction :: MonadIO m => String -> Shake.ShakeExtras -> ExceptT e IdeAction a -> ExceptT e m a
+runIdeAction _herald s i = ExceptT $ liftIO $ runReaderT (Shake.runIdeActionT $ runExceptT i) s
+
+-- | useE is useful to implement functions that aren’t rules but need shortcircuiting
+-- e.g. getDefinition.
+useWithStaleFast :: IdeRule k v => k -> NormalizedFilePath -> ExceptT GhcidePluginError IdeAction (v, PositionMapping)
+useWithStaleFast k = maybeToExceptT (RuleFailed k) . MaybeT . Shake.useWithStaleFast k
```

Comment:
I'm not totally sure why this gets a different error than `useWithStale` :thinking: More generally I think it would be good for the errors to match the function names. We can rename the functions... but let's keep the two consistent.